preflight_check/check.py

```diff
@@ -0,0 +1,254 @@
+        else:
+            task_inst.info("{} is available".format(timekeeper))
+            if crmshutils.service_is_enabled(timekeeper):
+                task_inst.info("{} is enabled".format(timekeeper))
+            else:
+                task_inst.warn("{} is disabled".format(timekeeper))
+
+            if crmshutils.service_is_active(timekeeper):
+                task_inst.info("{} is active".format(timekeeper))
+            else:
+                task_inst.warn("{} is not active".format(timekeeper))
+
+
+def check_watchdog():
+    """
+    Verify watchdog device
+    """
+    task_inst = task.TaskCheck("Checking watchdog")
+    with task_inst.run():
+        if not crmshutils.service_is_active("sbd.service") and not crmshboot.check_watchdog():
+            task_inst.warn("Watchdog device must be configured if want to use SBD!")
+
+
+def check_port_open(task, firewall_type):
+    """
+    Check whether corosync port is blocked by iptables
+    """
+    ports = utils.corosync_port_list()
+    if not ports:
+        task.error("Can not get corosync's port")
+        return
+
+    if firewall_type == "firewalld":
+        rc, out, err = crmshutils.get_stdout_stderr('firewall-cmd --list-port')
+        if rc != 0:
+            task.error(err)
+        for p in ports:
+            if re.search(' {}/udp'.format(p), out):
+                task.info("UDP port {} is opened in firewalld".format(p))
+            else:
+                task.error("UDP port {} should open in firewalld".format(p))
+    if firewall_type == "SuSEfirewall2":
+        #TODO
+        pass
+
+
+def check_firewall():
+    """
+    Check firewall
+    """
+    task_inst = task.TaskCheck("Checking firewall")
+    with task_inst.run():
+        for item in ("firewalld", "SuSEfirewall2"):
+            if crmshutils.package_is_installed(item):
+                task_inst.info("{}.service is available".format(item))
+                if crmshutils.service_is_active(item):
+                    task_inst.info("{}.service is active".format(item))
+                    check_port_open(task_inst, item)
+                else:
+                    task_inst.warn("{}.service is not active".format(item))
+                break
+        else:
+           task_inst.warn("Failed to detect firewall")
+
+
+def check_cluster():
+    """
+    A set of functions to check cluster state
+    """
+    print("\n============ Checking cluster state ============")
+    if not check_cluster_service():
+        return
+    check_fencing()
+    check_nodes()
+    check_resources()
+
+
+def check_cluster_service(quiet=False):
+    """
+    Check service status of pacemaker/corosync
+    """
+    task_inst = task.TaskCheck("Checking cluster service", quiet=quiet)
+    with task_inst.run():
+        for s in ("pacemaker", ):
+            if crmshutils.service_is_enabled(s):
+                task_inst.info("{} is enabled".format(s))
+            else:
+                task_inst.warn("{} is disabled".format(s))
+
```

Comment:
We also need to check "corosync" should be disabled.

Reply:
OK. Changed